### PR TITLE
add caddy server as system install

### LINKED ISSUE #1069: Add Caddy Server as System install
## Expected Behaviour

Add [caddy](https://github.com/caddyserver/caddy) server as system install, so user can install the same with `arkade system install caddy` given user has appropriate permission

### PR DESCRIPTION
**Pull request by nitishkumar71**
## Motivation and Context

- [x] I have raised an issue to propose this change, which has been given a label of `design/approved` by a maintainer ([required](https://github.com/alexellis/arkade/blob/master/CONTRIBUTING.md))
Closes #1069

## How Has This Been Tested?

Executed the command `arkade system install caddy` produced below output
```
Installing Caddy Server to /usr/local/bin
Installing version: v2.7.6 for: amd64
Downloading from: https://github.com/caddyserver/caddy/releases/download/v2.7.6/caddy_2.7.6_linux_amd64.tar.gz
13.98 MiB / 13.98 MiB [--------------------------------------------------------------------------------------------------------------------------------------------------------------] 100.00%
Downloaded to: /tmp/arkade-345065249/caddy_2.7.6_linux_amd64.tar.gz
Unpacking Caddy to: /usr/local/bin/caddy
```
If updating or adding a new CLI to `arkade get`, run:

```
go build && ./hack/test-tool.sh TOOL_NAME
```

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Documentation

- [ ] I have updated the list of tools in README.md if (required) with `./arkade get --format markdown`
- [ ] I have updated the list of apps in README.md if (required) with `./arkade install --help`

## Checklist:
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I've read the [CONTRIBUTION](https://github.com/alexellis/arkade/blob/master/CONTRIBUTING.md) guide
- [x] I have signed-off my commits with `git commit -s`

- [ ] I have tested this on arm, or have added code to prevent deployment

### CODE REVIEW

File: cmd/system/caddy.go [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+		Use:   "caddy",
+		Short: "Install Caddy Server",
+		Long:  `Install Caddy Server which is an extensible server platform that uses TLS by default`,
+		Example: `  arkade system install caddy
+  arkade system install caddy --version <version>`,
+		SilenceUsage: true,
+	}
+
+	command.Flags().StringP("version", "v", "", "The version or leave blank to determine the latest available version")
+	command.Flags().String("path", "/usr/local/bin", "Installation path, where a caddy subfolder will be created")
```
Comment (alexellis) [inline cmd/system/caddy.go:29]: It's not the usual pattern, but Caddy installs to /usr/bin/

Have a look at the faasd script to see what else may be missing, including adding users etc.

For adding users, we will need to detect the OS from /etc/os-release and reject the command if it's not debian or ubuntu.

In the future, if someone needs this for Alpine or Fedora, they can add a patch to run useradd with the correct syntax for their distro.

There are also some directories created and chmod is run.

**Review by alexellis — CHANGES_REQUESTED** [review]
Good start, but there's more needed for a system install to match how Caddy is installed with faasd: https://github.com/openfaas/faasd/blob/master/hack/install.sh

**Comment by alexellis** [comment]
The new changes look like what I was expecting.

Could you test this and see if Caddy starts up correctly on a new Linux VM you create somewhere?

File: cmd/system/caddy.go [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+		caddyUser := "caddy"
+		if _, err = user.Lookup(caddyUser); errors.Is(err, user.UnknownUserError(caddyUser)) {
+			addUserCmd := fmt.Sprintf("$SUDO useradd --system --home %s --shell /bin/false %s", caddyHomeDir, caddyUser)
+			shellCmds = append([]string{addUserCmd}, shellCmds...)
+
+			fmt.Printf("User created for caddy server.\n")
+		}
+
+		commands := strings.Join(shellCmds, "; ")
+		if err = exec.Command(commands).Run(); err != nil {
```
Comment (alexellis) [inline cmd/system/caddy.go:149]: Could you try go-execute which is used elsewhere? It has a shell option if needed.

File: cmd/system/caddy.go [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+
+		caddyHomeDir := "/var/lib/caddy"
+		os.MkdirAll(caddyHomeDir, fs.ModeDir)
+		caddyConfDir := "/etc/caddy"
+		os.MkdirAll(caddyConfDir, fs.ModeDir)
+		fmt.Printf("Created caddy home %s and Conf %s directory\n", caddyHomeDir, caddyConfDir)
+		// TODO: Add empty CaddyFile and if required FAAS_DOMAIN setup too
+
+		shellCmds := []string{
+			"$SUDO chown --recursive caddy:caddy /var/lib/caddy",
```
Comment (alexellis) [inline cmd/system/caddy.go:133]: I think $SUDO could be removed, and arkade could be run as `sudo arkade` if needed.

File: cmd/system/caddy.go [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+			"$SUDO chown --recursive caddy:caddy /etc/caddy",
+			"$SUDO systemctl enable caddy",
+			"$SUDO systemctl daemon-reload",
+			"$SUDO systemctl start caddy",
+		}
+
+		caddyUser := "caddy"
+		if _, err = user.Lookup(caddyUser); errors.Is(err, user.UnknownUserError(caddyUser)) {
+			addUserCmd := fmt.Sprintf("$SUDO useradd --system --home %s --shell /bin/false %s", caddyHomeDir, caddyUser)
```
Comment (alexellis) [inline cmd/system/caddy.go:142]: Try go-execute again which splits out cmd vs args

**Comment by alexellis** [comment]
@nitishkumar71 suggested using the "get" method from the tools code. I think it would be worth trying out as an experiment.

**Comment by nitishkumar71** [comment]
Re-executing system install produced below output
```
Installing Caddy Server to /usr/bin
Downloading: https://github.com/caddyserver/caddy/releases/download/v2.8.1/caddy_2.8.1_linux_amd64.tar.gz
13.92 MiB / 13.92 MiB [-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------] 100.00%
/tmp/arkade-2170843838/caddy_2.8.1_linux_amd64.tar.gz written.
2024/06/02 01:51:25 Extracted: /tmp/arkade-2170843838/caddy
2024/06/02 01:51:25 Copying /tmp/arkade-2170843838/caddy to /usr/bin/caddy
Downloaded caddy.service file to /tmp/arkade-2005628698/caddy.service
Copied caddy.service file to /etc/systemd/system/caddy.service
Created caddy home /var/lib/caddy and Conf /etc/caddy directory
User created for caddy server.
Created symlink /etc/systemd/system/multi-user.target.wants/caddy.service → /etc/systemd/system/caddy.service.
```

File: cmd/system/caddy.go (hunk 4 above)
Comment (nitishkumar71) [reply cmd/system/caddy.go]: Done

File: cmd/system/caddy.go (hunk 3 above)
Comment (nitishkumar71) [reply cmd/system/caddy.go]: Done

File: cmd/system/caddy.go (hunk 2 above)
Comment (nitishkumar71) [reply cmd/system/caddy.go]: Done

File: cmd/system/caddy.go [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+		version, _ := cmd.Flags().GetString("version")
+		progress, _ := cmd.Flags().GetBool("progress")
+
+		arch, osVer := env.GetClientArch()
+
+		if strings.ToLower(osVer) != "linux" {
+			return fmt.Errorf("this app only supports Linux")
+		}
+
+		tools := get.MakeTools()
```
Comment (alexellis) [inline cmd/system/caddy.go:50]: Does this change mean that "caddy" also appears via `arkade get caddy`?

**Review by alexellis — CHANGES_REQUESTED** [review]
Could you see the suggestions please?

File: cmd/system/caddy.go [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+		var tool *get.Tool
+		for _, t := range tools {
```
[the remark this answers is not in the document]
Comment (nitishkumar71) [reply cmd/system/caddy.go:52]: Sure. I would try to progressively apply this approach in system install apps.

File: cmd/system/caddy.go [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+
+		if _, err = executeShellCmd(context.Background(), "systemctl", "enable", "caddy"); err != nil {
+			return err
+		}
+
+		if _, err = executeShellCmd(context.Background(), "systemctl", "daemon-reload"); err != nil {
+			return err
+		}
+
+		if _, err = executeShellCmd(context.Background(), "systemctl", "start", "caddy"); err != nil {
```
[the remark this answers is not in the document]
Comment (nitishkumar71) [reply cmd/system/caddy.go]: Done

File: cmd/system/caddy.go (hunk 5 above)
Comment (nitishkumar71) [reply cmd/system/caddy.go:50]: Agree. will start working toward migrating system apps to use this new approach and enable apps which we want as system install only

**Review by alexellis — APPROVED** [review]
LGTM

**Pull request merged**